pkg/nodeagent/controller/operatingsystemconfig/reconciler.go

```diff
@@ -379,6 +379,10 @@ func (r *Reconciler) applyChangedUnits(ctx context.Context, log logr.Logger, uni
 
 func (r *Reconciler) removeDeletedUnits(ctx context.Context, log logr.Logger, node client.Object, units []extensionsv1alpha1.Unit) error {
 	for _, unit := range units {
+		// Assume that the unit has been created by gardener-node-agent if it has content.
+		// Otherwise, it might be a default OS unit which was enabled/disabled or where drop-ins were added.
+		unitCreatedByNodeAgent := unit.Content != nil
```

Comment:
This looks like a special solution for the units we know of as of today - but it doesn't work in general.
Wouldn't it be a better approach to augment https://github.com/gardener/gardener/blob/554451cd97e08099728fe9687e62e60ab6f9c88a/pkg/nodeagent/controller/operatingsystemconfig/changes.go#L59-L62 with a new field `newlyCreated bool` that is `true` only when the unit did not exist before?

Reply:
This would require changes in the way we store the state of OSC on the node. It would only be possible to determine `newlyCreated` on the very first time we touch the unit. e.g by checking whether the file exists or not. 
Thus, we have to add this field to the state and must not change it in subsequent updates.
If a unit was created by gardener via userdata, this solution would falsely classify the Unit as non-Gardener.

Btw. why does it not work in general? The `Content` of a `Unit` is its only field which is written to disk at the moment.

Reply:
> This would require changes in the way we store the state of OSC on the node. It would only be possible to determine `newlyCreated` on the very first time we touch the unit. e.g by checking whether the file exists or not. Thus, we have to add this field to the state and must not change it in subsequent updates.

Correct. Do you see a general problem with this approach?
In our implementation, I see that we just store the old OSC in disk and use it to compute the changes. This means we cannot easily augment it with a new field (we can, but it's somewhat cumbersome). Instead of having this `newlyCreated` field, GNA could just place another file next to the file it creates (`.<filename>.created-by-gardener-node-agent` or similar) to indicate that it was the creator. This can then be looked up when computing the OSC changes.

> If a unit was created by gardener via userdata, this solution would falsely classify the Unit as non-Gardener.

That's a good point. Generally, we shouldn't create units via user-data (except for GNA itself). I see that we also add the `user` and `sshd` unit. With above approach, we could also just write the additional file (`.<filename>.created-by-gardener-node-agent`) for these units in the user-data to prevent the wrong classification.

> Btw. why does it not work in general? The `Content` of a `Unit` is its only field which is written to disk at the moment.

If you think about the problem "GNA should not delete units it hasn't created" then the natural solution to this is "GNA must remember which units it has created and which units already existed before". You would never answer this with "GNA should not delete units without content in their unit files". You already state yourself that this assumption is fuzzy and optimistic in this comment:

```golang
// Assume that the unit has been created by gardener-node-agent if it has content.
// Otherwise, it might be a default OS unit which was enabled/disabled or where drop-ins were added.
```

All I'm saying is that this assumption might work for now/all the cases we know, but it's not a proper general solution. Unless there is a technical challenge with the proper solution that is hard/costly to overcome, I usually prefer to go for it. This typically pays off in the long run (we all will forget about this optimistic assumption and might hit a case where it doesn't work this way, eventually forcing us to come back to this and refactor it).

Reply:
> I don't know any other linux tool which remembers that it created a file by creating a second file in the same directory. In my opinion, this looks hacky. Instead, I prefer a single file in the GNAs config folder storing all filenames created by GNA. It could be a different file than the one for the OSC state to avoid additional complexity because of merging things.

Fair enough.

> `gardener-node-agent` only writes systemd units to the disk if [`Unit.Content != nil`](https://github.com/gardener/gardener/blob/592fec295f97db53260f270e29ff153b25441cb1/pkg/nodeagent/controller/operatingsystemconfig/reconciler.go#L310). Isn't it reasonable to use the same logic for the delete case? We could create a helper function to make this knowledge explicit.

Yet, this doesn't mean that GNA originally created the unit. It might have existed before already. Also, what if the OSC once had a unit with content, but at a later revision the content is removed while the unit remains (maybe with drop-ins)?
I don't have a real-world use case at hand, I'm just trying to explore the edges. If you still are not convinced and think the proposed heuristic is good enough, then please proceed. I won't stop you.

Reply:
> Yet, this doesn't mean that GNA originally created the unit. It might have existed before already. Also, what if the OSC once had a unit with content, but at a later revision the content is removed while the unit remains (maybe with drop-ins)?

Afaik we don't cover these cases atm. If a `Unit.Content` switches from non-nil to nil, GNA won't touch the corresponding `*. service` file and keeps it in its last non-nil state.
We also don't support restoring the content of overwritten files. This might be also a nice feature. We might not really need the restoring in the first place. However, operators would notify that GNA overwrote some files because GNA would need to store the previous file somewhere. This might help in future debugging cases e.g. if GNA will overwrite a default config with a non-working one. 

That's more than I anticipated, but I'll work on this. It also makes sense to create tests cases for the file handling. Such cases are very hard to debug on a running node.

Reply:
Okay, new approach 😉 
I implemented an own `afero.Fs` for gardener-node-agent which logs its operations and offers additional methods to delete only files/directories it created.
It is probably quite slow, because it writes its state into a yaml file. As node-agent OSC controller is not performing too many operations on the file-system this should be enough. Otherwise, there are more powerful options to store the state.

There are no tests for the new FS yet. It is meant to collect feedback if you like this approach.

Reply:
Thanks :) I think the approach is fine. It needs some polishing and improved error messages, etc., but conceptually this is what we need. The impact on the business logic looks minimal. Cool!

> It is probably quite slow when there many file operations, because it writes its state into a yaml file. As node-agent OSC controller is not performing too many operations on the file-system this should be enough. Otherwise, there are more powerful options to store the state.

What do you have in mind when you say "more powerful options to store the state"? The only real scenario where GNA really writes "many" files is the initial reconciliation, because everything is new here, right? In subsequent reconciliations, the number of units/files that change simultaneously is rather small.

Reply:
You are right, the number would be quite small. It writes its state only when it touches a file for the first time but not if it changes it for a second or third time.
We usually change units/files and rarely create new ones. Thus, in subsequent reconciliations it will save its state only for the temporary files we use to apply the changes. 

"more powerful" would be something like sqlite for storing the state, but it is also more difficult. I don't think we need that, but I just wanted to mention that this FS should not be used to copy many small files at a large scale.